content/blog/2024/2024-05-15-java-functional-interfaces.md

```diff
@@ -0,0 +1,2028 @@
+| Reference to an Instance Method of a Particular Object  | Refers to an instance method of a specific object. The instance must be explicitly defined before using the method reference. | <code>Supplier<String> getMessage = stringUtils::getMessage;</code>|
+| Reference to an Instance Method of an Arbitrary Object of a Particular Type | Refers to an instance method of an arbitrary object of a specific type. We commonly use this type in stream operations, where Java determines the object type at runtime. | <code>List<String> uppercasedWords = words.stream()<br>.map(String::toUpperCase)<br>.collect(Collectors.toList());<code>|
+| Reference to a Constructor                              | Refers to a class constructor, allowing us to create new instances. This type is useful when we need to create objects without explicitly calling a constructor. | <code>Supplier<Car> carSupplier = Car::new;<code>|
+
+## Predicates
+
+**Predicates are functional interfaces in Java that represent boolean-valued functions of a single argument.** They are commonly used for filtering, testing, and conditional operations.
+
+The `Predicate` functional interface is part of the `java.util.function` package and defines a functional method `test(T t)` that returns a `boolean`.  It also provides default methods that allow combine two predicates.
+
+```java
+@FunctionalInterface
+public interface Predicate<T> {
+    boolean test(T t);
+    // default methods
+}
+```
+
+The `test()` method evaluates the predicate on the input argument and determines whether it satisfies the condition defined by the predicate.
+
+We often use predicates with the `stream()` API for filtering elements based on certain conditions. Pass them as arguments to methods like `filter()` to specify the criteria for selecting elements from a collection.
+
+Let's see filtering in action:
+
+```java
+public class PredicateTest {
+  @Test
+  void testFiltering() {
+    List<Integer> numbers = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+    Predicate<Integer> isEven = num -> num % 2 == 0;
+    
+    List<Integer> actual = numbers.stream().filter(isEven).toList();
+    List<Integer> expected = List.of(2, 4, 6, 8, 10);
+    
+    Assertions.assertEquals(expected, actual);
+  }
+}
+```
+
+In the test `testFiltering()` method, first we populate a list of integers. Then we define a predicate `isEven` to check if a number is even. Using `stream()` and `filter()` methods, we filter the list to contain only even numbers. Finally, we compare the filtered list to expected list.
+
+### Combining Predicates
+
+**We can combine predicates using logical operators such as `and()`, `or()`, `negate()` and `not()` to create complex conditions.**
+
+Let's see how to combine the practices:
```

Comment:
It would be interesting to elaborate on the order in which the predicates are evaluated. Consider the following example:

```
List<Integer> numbers = List.of(-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5);
Predicate<Integer> isPositive = num -> num > 0;
Predicate<Integer> isOdd = num -> num % 2 == 1;
Predicate<Integer> isDiv3 = num -> num % 3 == 0;

Predicate<Integer> test1 = isPositive.or(isDiv3).and(isOdd);
numbers.stream().filter(test1).toList().forEach(System.out::println);

Predicate<Integer> test2 = isPositive.and(isOdd).or(isDiv3);
numbers.stream().filter(test2).toList().forEach(System.out::println);

Predicate<Integer> test3 = (isPositive.and(isOdd)).or(isDiv3);
numbers.stream().filter(test3).toList().forEach(System.out::println);

Predicate<Integer> test4 = isPositive.and((isOdd).or(isDiv3));
numbers.stream().filter(test4).toList().forEach(System.out::println);
```

I know that this isn't easy to describe, but maybe we can find a good way to do that.